File: 2022/deno/15.ts
type Coord = [x: number, y: number];

type Sensor = {
  center: Coord;
  beacon: Coord;
  radius: number;
};

/**
 * Helpers
 */

const merge = (intervals: Coord[]) => {
  if (intervals.length < 2) return intervals;

  const result = [];
  let [previous] = intervals;

  for (const interval of intervals) {
    if (previous[1] >= interval[0]) {
      previous = [previous[0], Math.max(previous[1], interval[1])];
    } else {
      result.push(previous);
      previous = interval;
    }
  }

  result.push(previous);

  return result;
};

const intersection = ({
  center,
  radius: r,
  y,
}: {
  center: number[];
  radius: number;
  y: number;
}): [] | Coord => {
  const [cx, cy] = center;

  if (y > cy + r || y < cy - r) return [];

  // r = |x - cx| + |y - cy|
  // |x - cx| = r - |y - cy| = delta
  // x = cx + delta; cx - delta
  const delta = r - Math.abs(y - cy);

  // overlaps [from, to]
  return [cx - delta, cx + delta];
};

const coverage = (y: number, grid: Sensor[]) =>
  merge(
    grid
      .map(({ center, radius }) => intersection({ center, radius, y }))
      .filter((overlap): overlap is Coord => Boolean(overlap.length))
      .sort((a, b) => a[0] - b[0])
  );

const beaconsAtOffset = (offset: number, grid: Sensor[]) =>
  new Set(
    grid
      .filter(({ beacon }) => beacon[1] === offset)
      .map(({ beacon }) => `${beacon}`)
  ).size;

const outOfSensorRange = ([x, y]: Coord, grid: Sensor[]) =>
  grid.every(({ center, radius }) => {
    const [cx, cy] = center;

    const distance = Math.abs(cx - x) + Math.abs(cy - y);

    return distance >= radius;
  });

const solve = async (example = false) => {
  /**
   * Part One
   */
  const input = await Deno.readTextFile(
    `./input/${example ? "example" : "15"}.in`
  );

  const data = input.split("\n");

  const sensors = data.map<Sensor>((row) => {
    const [xS, yS, xB, yB] = row
      .replace(":", "")
      .replaceAll(",", "")
      .replaceAll("=", " ")
      .split(" ")
      .map(Number)
      .filter(isFinite);

    const radius = Math.abs(xB - xS) + Math.abs(yB - yS);
    return { center: [xS, yS], beacon: [xB, yB], radius };
  });

  const offset = example ? 10 : 2_000_000;
  const occupied = beaconsAtOffset(offset, sensors);
  const forbidden = coverage(offset, sensors)
    .map(([x0, x1]) => x1 - x0 + 1)
    .reduce((a, b) => a + b);

  console.log("Part one:", forbidden - occupied);

  /**
   * Part Two
   */

  //   const maxRange = example ? 20 : 4_000_000;

  const tuningFreq = ([x, y]: Coord) => x * 4_000_000 + y;

  // naive approach, takes about 2 seconds
  //   for (let y = 0; y < maxRange; y++) {
  //     const zones = coverage(y, sensors);

  //     if (zones.length === 2) {
  //       const [[, x0], [x1]] = zones;
  //       const x = Math.floor((x0 + x1) / 2); // or just x0 + 1

  //       console.log("Part two:", tuningFreq([x, y]));

  //       break;
  //     }
  //   }

  /**
   * Part two super fast!! Blazing fast!!!
   */

  // 45 degrees clockwise
  const forwards = ([x, y]: Coord): Coord => [x + y, y - x];
  // 45 degrees anti-clockwise - undoing the above
  const backwards = ([x, y]: Coord): Coord =>
    [(x - y) / 2, (x + y) / 2].map((x) => Math.floor(x)) as Coord;

  type Range = { from: number; to: number };
  type Projection = {
    x: Range;
    y: Range;
  };

  /**
   *
   * Rotating a Manhattan circle resolves
   * to a square. With a square it is easier to measure
   * gaps relative to other squares.
   *
   * The solution beacon has to live in between
   * squares separated by at least 2 units
   *
   *  x    x+2
   *  | - |
   *    x+1 -> beacon.x
   *
   * Same applies for Y axis.
   *
   * Once X and Y in the rotated system are found,
   * rotate back and eliminate those which
   * are part of a sensor's coverage.
   *
   * There'll be only one left.
   *
   */

  // rotates 45 degrees to turn the sensor areas from
  // manhattan circles into squares
  // returns the coordinates of the corners
  const squaredSensors = sensors.map(({ center, radius }) => {
    const [cx, cy] = center;
    const corners: [Coord, Coord, Coord, Coord] = [
      [cx - radius, cy],
      [cx, cy + radius],
      [cx + radius, cy],
      [cx, cy - radius],
    ];

    return corners.map(forwards);
  });

  // represent squares as their projections
  // into the x-axis: x0->x1 and y-axis: y0->y1
  const projections = squaredSensors.map<Projection>((corners) => {
    const [c0, c1, , c3] = corners;

    return {
      x: { from: c0[0], to: c1[0] },
      y: { from: c3[1], to: c0[1] },
    };
  });

  const xProjections = projections.map(({ x }) => x);
  const yProjections = projections.map(({ y }) => y);

  const hasValidGap = (
    pair: Readonly<[Range, Range | undefined]>
  ): pair is [Range, Range] => Boolean(pair[1]);

  const findGap = (self: Range, others: Range[]) =>
    others.find((other) => other !== self && other.from === self.to + 2);

  const xGaps: number[] = xProjections
    .map((self) => [self, findGap(self, xProjections)] as const)
    .filter(hasValidGap)
    .map(([self]) => self.to + 1);

  const yGaps = yProjections
    .map((self) => [self, findGap(self, yProjections)] as const)
    .filter(hasValidGap)
    .map(([self]) => self.to + 1);

  const [beacon, ...rest] = xGaps
    .flatMap((rx) => yGaps.map<Coord>((ry) => [rx, ry]))
    // undo the clockwise rotation
    .map(backwards)
    .filter((coord) => outOfSensorRange(coord, sensors));

  console.assert(rest.length === 0, "Found many beacons!");

  console.log("Part two:", tuningFreq(beacon));
};

solve();

File: 2022/swift/15.swift
import Foundation

let filename = "./input/15.in"

func merge(_ intervals: [[Int]]) -> [[Int]] {
    if intervals.count < 2 {
        return intervals
    }

    var result = [[Int]]()

    var previous = intervals[0]

    for interval in intervals {
        if previous[1] >= interval[0] {
            previous = [previous[0], max(previous[1], interval[1])]
        } else {
            result.append(previous)
            previous = interval
        }
    }

    result.append(previous)

    return result
}

func intersection(_ sensor: [Int], _ y: Int) -> [Int]? {
    if y > sensor[1] + sensor[4] || y < sensor[1] - sensor[4] {
        return nil
    }

    let delta = sensor[4] - abs(sensor[1] - y)

    return [sensor[0] - delta, sensor[0] + delta]
}

func beaconsAtOffset(offset: Int, sensors: [[Int]]) -> Int {
    Set(sensors.filter { $0[3] == offset }.map { [$0[2], $0[3]] }).count
}

func coverage(offset: Int, sensors: [[Int]]) -> [[Int]] {
    let asSegments = sensors.compactMap { intersection($0, offset) }
        .sorted { $0[0] < $1[0] }

    return merge(asSegments)
}

func rotateFwd(coord: [Int]) -> [Int] {
    return [coord[0] + coord[1], coord[1] - coord[0]]
}

func rotateBwd(coord: [Int]) -> [Int] {
    return [(coord[0] - coord[1]) / 2, (coord[0] + coord[1]) / 2]
}

func outOfSensorRange(_ coord: [Int], _ sensors: [[Int]]) -> Bool {
    sensors.allSatisfy {
        sensor in
        let cx = sensor[0]
        let cy = sensor[1]
        let radius = sensor[4]

        let distance = abs(cx - coord[0]) + abs(cy - coord[1])

        return distance >= radius
    }
}

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let sensors = contents.split(separator: "\n").map { row in

            var spec = String(row)
                .replacingOccurrences(of: ",", with: "")
                .replacingOccurrences(of: "=", with: " ")
                .replacingOccurrences(of: ":", with: "")
                .split(separator: " ")
                .compactMap { Int(String($0)) }

            let radius = abs(spec[0] - spec[2]) + abs(spec[1] - spec[3])

            spec.append(radius)

            return spec
        }

        let offset = 2_000_000

        let occupied = beaconsAtOffset(offset: offset, sensors: sensors)
        let forbidden = coverage(offset: offset, sensors: sensors)
            .map { $0[1] - $0[0] + 1 }
            .reduce(0, +)

        print("Part one:", forbidden - occupied)

        let maxRange = 4_000_000

        let tunningFreq: ([Int]) -> Int = {
            sensor in
            sensor[0] * maxRange + sensor[1]
        }

        let squared = sensors.map {
            sensor in
            let cx = sensor[0]
            let cy = sensor[1]
            let radius = sensor[4]

            return [
                [cx - radius, cy],
                [cx, cy + radius],
                [cx + radius, cy],
                [cx, cy - radius],
            ].map(rotateFwd)
        }

        let projections = squared.map {
            corners in
            let c0 = corners[0]
            let c1 = corners[1]
            let c3 = corners[3]

            return (x: (from: c0[0], to: c1[0]), y: (from: c3[1], to: c0[1]))
        }

        let xProjections = projections.map { $0.x }
        let yProjections = projections.map { $0.y }

        let xGaps = xProjections
            .map {
                segment in
                (segment, xProjections.first {
                    other in
                    segment != other && segment.to + 2 == other.from
                })
            }
            .filter { $0.1 != nil }
            .map { $0.0.to + 1 }

        let yGaps = yProjections
            .map {
                segment in
                (segment, yProjections.first {
                    other in
                    segment != other && segment.to + 2 == other.from
                })
            }
            .filter { $0.1 != nil }
            .map { $0.0.to + 1 }

        let beacons = xGaps.flatMap {
            x in
            yGaps.map {
                y in
                [x, y]
            }
        }.map(rotateBwd)
            .filter { outOfSensorRange($0, sensors) }

        assert(beacons.count == 1, "Found more than one possible beacon")

        print("Part two:", tunningFreq(beacons[0]))

    } catch {
        print(error)
    }
}

main()
